crates/mptnew/src/trie.rs

```diff
@@ -0,0 +1,809 @@
+use std::cell::RefCell;
+
+use alloy_rlp::Encodable;
+use bumpalo::Bump;
+use bytes::Buf;
+use revm_primitives::{keccak256, B256};
+use smallvec::SmallVec;
+
+use crate::{
+    bump_bufmut::BumpBytesMut,
+    hp::{
+        encoded_path_eq_nibs, encoded_path_strip_prefix, lcp, prefix_to_nibs, to_encoded_path,
+        to_nibs,
+    },
+    node::{NodeData, NodeId, NodeRef},
+};
+
+/// Initial capacity of [`MptTrie`]'s `rlp_scratch`.
+const RLP_SCRATCH_INIT_CAPACITY: usize = 600;
+
+const VALUE_RLP_BUFFER_CAPACITY: usize = 200;
+
+/// Sentinel index representing the null node when decoding and in internal references.
+/// In a default MPT, `nodes[0]` starts as `Null`, but the root may later be changed to a
+/// non-null node (e.g. `Digest`) for convenience. `NULL_NODE_ID` is still used by the decoder
+/// as the canonical "no node" identifier.
+const NULL_NODE_ID: NodeId = 0;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Node reference does not match the expected node reference in the parent.
+    #[error("node reference mismatch")]
+    NodeRefMismatch,
+    /// Triggered when an operation reaches an unresolved node. The associated `B256`
+    /// value provides details about the unresolved node.
+    #[error("reached an unresolved node: {0:#}")]
+    NodeNotResolved(B256),
+    /// Represents errors related to the RLP encoding and decoding.
+    #[error("rlp decode error: {0}")]
+    RlpError(#[from] alloy_rlp::Error),
+    /// Occurs when a value is unexpectedly found in a branch node.
+    #[error("branch node with value")]
+    ValueInBranch,
+}
+
+/// Arena-based implementation that stores all nodes in a flat vector and uses indices for better
+/// memory layout and performance. The lifetime parameter `'a` allows zero-copy deserialization by
+/// borrowing from the input buffer.
+#[derive(Debug, Clone)]
+pub struct MptTrie<'a> {
+    root_id: NodeId,
+
+    /// List of MPT nodes.
+    nodes: Vec<NodeData<'a>>,
+
+    /// Cache. Hashing/encoding often needs "what would this node look like in its parent"
+    cached_references: Vec<RefCell<Option<NodeRef<'a>>>>,
```

Comment:
btw even in Valery's PR, I wondered if we could get away using raw pointers for a little gain